target/arc/semfunc-v2.c

```diff
@@ -6902,517 +6902,350 @@ arc_gen_DMB (DisasCtxt *ctx, TCGv a)
+ *       src1 = src1 + src2;
  * }
  */
 
 int
 arc_gen_STD(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
 {
-    int ret = DISAS_NEXT;
-    int AA;
-    int ZZ;
-    TCGv address = tcg_temp_local_new();
-    TCGv temp_3 = tcg_temp_local_new();
-    TCGv temp_4 = tcg_temp_local_new();
-    TCGv pair = NULL;
-    TCGv temp_5 = tcg_temp_local_new();
-    TCGv tmp = tcg_temp_local_new();
-    TCGv temp_7 = tcg_temp_local_new();
-    TCGv temp_6 = tcg_temp_local_new();
-    TCGv temp_1 = tcg_temp_local_new();
-    TCGv temp_2 = tcg_temp_local_new();
-    TCGv temp_8 = tcg_temp_local_new();
-    AA = getAAFlag ();
-    ZZ = getZZFlag ();
-    tcg_gen_movi_tl(address, 0);
-    if (((AA == 0) || (AA == 1))) {
-        tcg_gen_add_tl(address, src1, src2);
-    }
-    if ((AA == 2)) {
-        tcg_gen_mov_tl(address, src1);
-    }
-    if (((AA == 3) && ((ZZ == 0) || (ZZ == 3)))) {
-        tcg_gen_shli_tl(temp_3, src2, 2);
-        tcg_gen_add_tl(address, src1, temp_3);
+    int aa = ctx->insn.aa;
+    int zz = ctx->insn.zz;
+
+    TCGv address = tcg_temp_new();
+    TCGv scaled_disp = tcg_temp_new();
+
+    /* compute destination address */
+    tcg_gen_mov_tl(address, src1);
+    if (aa == 0 || aa == 1) {
+        tcg_gen_add_tl(address, address, src2);
+    } else if (aa == 3) {
+        if (zz == 0 || zz == 3) {
+            tcg_gen_shli_tl(scaled_disp, src2, 2);
+        } else if (zz == 2) {
+            tcg_gen_shli_tl(scaled_disp, src2, 1);
+        }
+        tcg_gen_add_tl(address, src1, scaled_disp);
     }
-    if (((AA == 3) && (ZZ == 2))) {
-        tcg_gen_shli_tl(temp_4, src2, 1);
-        tcg_gen_add_tl(address, src1, temp_4);
-    }
-    setMemory(address, LONG, dest);
-    if (instructionHasRegisterOperandIn (0)) {
-        pair = nextReg (dest);
-        tcg_gen_addi_tl(temp_5, address, 4);
-        setMemory(temp_5, LONG, pair);
+
+    /* write the low 32 bits */
+    tcg_gen_qemu_st_tl(dest, address, ctx->mem_idx, MO_UL);
+
+    /* depending on the type of the destination operand, either ... */
+    if (ctx->insn.operands[0].type & ARC_OPERAND_IR) {
+        /* write the contents of the high 32 bits if it's a register, or ... */
+        tcg_gen_addi_tl(address, address, 4);
+        tcg_gen_qemu_st_tl(nextReg(dest), address, ctx->mem_idx, MO_UL);
     } else {
-        tcg_gen_movi_tl(tmp, 0);
-        TCGLabel *done_1 = gen_new_label();
-        tcg_gen_movi_tl(temp_7, 31);
-        getBit(temp_6, dest, temp_7);
-        tcg_gen_setcondi_tl(TCG_COND_EQ, temp_1, temp_6, 1);
-        tcg_gen_xori_tl(temp_2, temp_1, 1);
-        tcg_gen_andi_tl(temp_2, temp_2, 1);
-        tcg_gen_brcond_tl(TCG_COND_EQ, temp_2, arc_true, done_1);
-        tcg_gen_movi_tl(tmp, 4294967295);
-        gen_set_label(done_1);
-        tcg_gen_addi_tl(temp_8, address, 4);
-        setMemory(temp_8, LONG, tmp);
+        /* sign-extend to 64 bits if it's an immediate */
+        TCGv value_hi = tcg_temp_new();
+        tcg_gen_sari_tl(value_hi, dest, 31);
```

Comment:
Impact: minor

Here we're making an assumption about the length by shifting "31" bits right. Use `*_32` instead of `*_tl`.

```
tcg_gen_sari_32(value_hi, dest, 31)
```